### PR TITLE
FastQC: Only add the prefix if the file doesn't already start with it

### PR DESCRIPTION
**Pull request by ewels**
Follow on from https://github.com/nf-core/modules/pull/2282

Fix from @mahesh-panchal to avoid adding a prefix if the file already starts with that prefix (which it often does, as the prefix defaults to `meta.id`). So avoids filenames like `SRR389222_sub1_T1_SRR389222_sub1`.

See discussion [on Slack](https://nfcore.slack.com/archives/CJRH30T6V/p1667557919607909).

Needed for the nf-core/methylseq release: https://github.com/nf-core/methylseq/pull/251

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/fastqc/main.nf [hunk 1]
```diff
@@ -22,7 +22,7 @@ process FASTQC {
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    printf "%s\\n" $reads | while read f; do ln -s \$f ${prefix}_\$(basename \$f) ; done
+    printf "%s\\n" $reads | while read f; do [ -f ${prefix}_1.fastq.gz ] || ln -s \$f ${prefix}_\$(basename \$f) ; done
```
Comment (mahesh-panchal) [inline modules/nf-core/fastqc/main.nf:25]:
```suggestion
    printf "%s\\n" $reads | while read f; do [[ \$f =~ ^${prefix}.* ]] || ln -s \$f ${prefix}_\$(basename \$f) ; done
```

**Comment by ewels** [comment]
Needs testing to confirm the fix.

**Comment by ewels** [comment]
Will still need to fix md5s I guess, as file paths has changed and I bet this is reported inside the file contents.

**Comment by ewels** [comment]
I can't figure out these conda failures now. When I run the tests on GitPod with conda the files seem to be there, no idea why it's failing in CI.

**Comment by mahesh-panchal** [comment]
It's literally that issue with gzip I mentioned in Slack. https://nfcore.slack.com/archives/CJRH30T6V/p1667211637507889?thread_ts=1667201233.719109&cid=CJRH30T6V

**Comment by mahesh-panchal** [comment]
>  N E X T F L O W  ~  version 22.10.1
   2   │ Launching `./tests/subworkflows/nf-core/fastq_fastqc_umitools_trimgalore/main.nf` [cheesy_kimura] DSL2 - revision: ff5d741455
   3   │ WARN: The operator `first` is useless when applied to a value channel which returns a single value by definition
   4   │ WARN: The operator `first` is useless when applied to a value channel which returns a single value by definition
   5   │ WARN: The operator `first` is useless when applied to a value channel which returns a single value by definition
   6   │ Creating env using conda: bioconda::umi_tools=1.1.2 [cache /home/runner/pytest_workflow_z2qqi6l_/fastq_fastqc_umitools_trimgalore_test_fastq_fastqc_umitools_trimgalore_paired/work/conda/env-57ed1401cdde1c224250a4d54f05f4ba]
   7   │ Creating env using conda: bioconda::fastqc=0.11.9 [cache /home/runner/pytest_workflow_z2qqi6l_/fastq_fastqc_umitools_trimgalore_test_fastq_fastqc_umitools_trimgalore_paired/work/conda/env-e92705f017a3fab1fca3da28471361dd]
   8   │ [06/996118] Submitted process > test_fastq_fastqc_umitools_trimgalore_paired:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (test)
   9   │ Error executing process > 'test_fastq_fastqc_umitools_trimgalore_paired:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (test)'
  10   │ 
  11   │ Caused by:
  12   │   Missing output file(s) `*.html` expected by process `test_fastq_fastqc_umitools_trimgalore_paired:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (test)`
  13   │ 
  14   │ Command executed:
  15   │ 
  16   │   printf "%s\n" 1/test_1.fastq.gz 2/test_2.fastq.gz | while read f; do [[ $(basename $f) =~ ^test.* ]] && ln -s $f $(basename $f) || ln -s $f test_$(basename $f) ; done
  17   │   fastqc  --threads 2 test*
  18   │   
  19   │   cat <<-END_VERSIONS > versions.yml
  20   │   "test_fastq_fastqc_umitools_trimgalore_paired:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC":
  21   │       fastqc: $( fastqc --version | sed -e "s/FastQC v//g" )
  22   │   END_VERSIONS
  23   │ 
  24   │ Command exit status:
  25   │   0
  26   │ 
  27   │ Command output:
  28   │   (empty)
  29   │ 
  30   │ Command error:
  31   │   Skipping 'test_1.fastq.gz' which didn't exist, or couldn't be read
  32   │   Skipping 'test_2.fastq.gz' which didn't exist, or couldn't be read

There's an odd `1/` and `2/` in the name.

File: modules/nf-core/fastqc/main.nf [hunk 2]
```diff
@@ -22,8 +22,8 @@ process FASTQC {
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    printf "%s\\n" $reads | while read f; do ln -s \$f ${prefix}_\$(basename \$f) ; done
-    fastqc $args --threads $task.cpus ${prefix}_*
+    printf "%s\\n" $reads | while read f; do [[ \$(basename \$f) =~ ^${prefix}.* ]] && ln -s \$f \$(basename \$f) || ln -s \$f ${prefix}_\$(basename \$f) ; done
```
Comment (mahesh-panchal) [inline modules/nf-core/fastqc/main.nf:25]:
```bash
&& ln -s \$f \$(basename \$f)
```
Why is this here? It's reversing the `stageAs`.

Comment (ewels) [reply modules/nf-core/fastqc/main.nf:25]: See also your [other comment](https://github.com/nf-core/modules/pull/2465#issuecomment-1305251788):

> There's an odd `1/` and `2/` in the name.

This is where that numeric prefix directory is coming from.

Comment (mahesh-panchal) [reply modules/nf-core/fastqc/main.nf:25]: Ah, I see. That's a good point about not including original files, but if we remove the `stageAs` then we can simplify this expression, and the `fastqc ${prefix}*` ( already there ) should exclude any files we don't want to run on. The `stageAs` is normally used to prevent filename collisions, so if we leave it in, we need to make sure the symlinking doesn't also reintroduce this issue.

**Comment by mahesh-panchal** [comment]
Tried testing whether removing the `stageAs` and remaking the test.yml s is a solution, but the `nf-core subworkflows create-test-yml` is breaking on me: https://github.com/nf-core/tools/issues/2003

**Comment by ewels** [comment]
Yeah, I was just doing all the md5s etc by hand last night 👀

**Comment by ewels** [comment]
ok @mahesh-panchal - removed the `stageAs` and simplified the bash logic quite a bit. Updated all of the md5s accordingly.

**Review by mahesh-panchal — APPROVED** [review]
LGTM 🚀

**Comment by mahesh-panchal** [comment]
Why are the tests suddenly off after all going through?

**Comment by ewels** [comment]
ok, as solved on Slack: The HTML reports contain the _date_ but not the _time_. So the md5sums were static during a single working day but would break again the day after 🤦🏻 

Now replaced all of the md5sum tests with a `containers` string arbitrarily chosen from the HTML.
